src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,183 @@
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly mtd: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.mtd.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+            console.log(`Loop iteration with sleep time: ${sleepTime}ms`);
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
+            const workflowUUID = `sched-${this.mtd.name}-${nextExecTime.toISOString()}`;
+            this.dbosExec.logger.debug(`Executing scheduled workflow ${workflowUUID}`);
+            const wfParams = { workflowUUID: workflowUUID };
+            // All operations annotated with Scheduled decorators must take in these four
+            const args: ScheduledArgs = [nextExecTime, new Date(), 0, 0]; // TODO calculate outstanding numbers
+
+            // We can only guarantee exactly-once-per-message execution of transactions and workflows.
+            if (this.mtd.workflowConfig) {
+                // Execute the transaction
+                await this.dbosExec.workflow(this.mtd.registeredFunction as Workflow<[Date, Date, number, number], unknown>, wfParams, ...args);
+            }
+            else {
+                this.dbosExec.logger.error(`Function ${this.mtd.name} is @scheduled but not a workflow`);
+            }
+
+            // Record the time of the wf kicked off
+            const dbTime = await this.dbosExec.systemDatabase.setLastScheduledTime(this.mtd.name, nextExecTime.getTime());
+            if (dbTime && dbTime > nextExecTime.getTime()) nextExecTime.setTime(dbTime);
+            this.lastExec = nextExecTime;
```

Comment:
Interestingly this "adjust" to clock skew across nodes. Is there another reason to adjust the interval's exec time?

Reply:
It doesn't.  This is a logical time, not a clock time, as I mentioned.  If there is skew, one machine will be winning the race to execute the wf.

Reply:
I think (?) I got it: given what we discussed in the previous comment, it seems that `setLastScheduledTime` should always be the same across nodes. Why are we even resolving conflicts there then?

Reply:
Must be what Qian is saying: another node scheduled the task in the "future"

Reply:
If another VM has gone ahead and started a workflow and bumped the DB forward, why would we start that one in the next loop iteration?  We wouldn't.  We're going to the DB anyway, may as well bring back the latest record value.